new ad create enhanced tool to validate and create multiple detectors

amitgalitz opened the pull request:
### Description



This PR introduces CreateAnomalyDetectorToolEnhanced, a tool that leverages LLM capabilities to automatically generate, validate, and create anomaly detectors for different given indices

## Implementation Overview
The tool implements a step-by-step processing flow that handles batch operations on indices while maintaining robust error handling and validation mechanisms. 

### Code Flow
1. Extract & validate indices →
2. Get index mappings →
3. LLM generates detector config →
4. Validate detector structure →
5. Optimize parameters (suggest API) →
6. Validate model requirements →
7. Create & start detector

The implementation includes utilizing an LLM configuration generation, and retrying on different stages with the validate and suggest api to get optimal configuration that will lead to detector initializing
detector reliability.

The tool processes indices sequentially to prevent LLM throttling and implements retry logic with LLM-assisted error correction. Each index undergoes mapping analysis, LLM configuration generation, detector validation, suggest api based changes, and model validation before final deployment.

## Usage Example

Input:
```
{
  "indices": ["ecommerce-data", "server-logs"]
}
```

Output:
```
{
  "ecommerce-data": {
    "status": "success",
    "detectorId": "abc123",
    "detectorName": "ecommerce-data-detector-xyz",
    "createResponse": "Detector created successfully",
    "startResponse": "Detector started successfully"
  },
  "server-logs": {
    "status": "failed_validation",
    "error": "Insufficient data for model training"
  }
}
```

## Processing Flow
The implementation follows a step-by-step flow for each index: extraction and validation, mapping retrieval, LLM configuration generation, detector structure validation, parameter optimization via suggest API, model requirement validation, and finally detector creation and activation. The tool maintains state consistency across batch operations and writes it to the detector result object that is returned as a string so an LLM can use this as one of its tools

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
<!-- List any other related issues here -->

### Check List
- [x] New functionality includes testing.
- [x] Commits are signed per the DCO using `--signoff`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/skills/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

amitgalitz committed:
new ad create enhanced tool to validate and create multiple detectors

Signed-off-by: Amit Galitzky <amgalitz@amazon.com>

dbwiddis commented on src/main/java/org/opensearch/agent/tools/utils/AnomalyDetectorToolHelper.java:
```diff
@@ -0,0 +1,68 @@
+            for (String index : indices) {
+                if (index.startsWith(".")) {
+                    throw new IllegalArgumentException("System indices not supported: " + index);
+                }
```
I think we have [better detection of System indices](https://github.com/opensearch-project/OpenSearch/blob/f76826cc034f652a4cd99c51046a51f1ed6d0512/server/src/main/java/org/opensearch/indices/SystemIndices.java#L66-L82) than the leading dot.  [Example usage](https://github.com/opensearch-project/OpenSearch/blob/f76826cc034f652a4cd99c51046a51f1ed6d0512/server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java#L448).

dbwiddis commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+            log.error("Unknown response format. Available keys: {}", dataAsMap.keySet());
+            return null;
+        }
+    }
```
This hard-coded parsing and relying on exception handling seems like it could be improved.  At least add a comment showing the expected format, so if this breaks in the future, the next maintainer (or future-you) can quickly identify the format change without walking through the code.

dbwiddis commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+        String fixPrompt = createFixPrompt(originalSuggestions, validationError);
+        String fullPrompt = contextBuilder.toString() + fixPrompt;
+
+        callLLM(fullPrompt, tenantId, ActionListener.wrap(fixedResponse -> {
```
Lots of duplicated code between `retryDetectorValidation()` and `retryModelValidation`, consider a helper method for the common parts.

dbwiddis commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+            }
+
+            String finalResponse = extractResponseFromDataAsMap(dataAsMap);
+            if (Strings.isNullOrEmpty(finalResponse)) {
```
I like this, you should use it everwhere there's a `foo == null || foo.isEmpty()`

dbwiddis commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+        "Enhanced tool for creating anomaly detector configurations. Takes an index name, extracts the index mappings, and uses LLM to generate complete detector JSON configurations ready for the create detector API.";
+
+    private static final String EXTRACT_INFORMATION_REGEX =
+        "(?s).*\\{category_field=([^|]*)\\|aggregation_field=([^|]*)\\|aggregation_method=([^|]*)\\|interval=([^}]*)}.*";
```
This seems to be a really strict regex that would fail on things like whitespace or the keys being in a different order (and cause retries which lose the benefit of the optimal date field.).  

Given we're just extracting characters between known strings, it seems there's likely a better way of iterating between the `|` characters, splitting on `=`, and trimming the key/value into a map?

dbwiddis commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+                                        "Detector validation failed after " + MAX_DETECTOR_VALIDATION_RETRIES + " retries: " + errorMessage
+                                    )
+                                );
+                        }
```
This conditional (`currentRetry >= MAX_DETECTOR_VALIDATION_RETRIES`) should be a fast-fail at the top of the `validateDetectorPhase` method, otherwise you spend a lot of time validating the detector which, if validation succeeds, you suddenly can't use.

dbwiddis commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+
+                if (response.getIssue() != null) {
+                    String issueAspect = response.getIssue().getAspect().toString();
+                    boolean isBlockingError = issueAspect != null && issueAspect.toLowerCase(Locale.ROOT).startsWith("detector");
```
Is starting with "detector" always blocking?

Is there some helper method somewhere which defines blocking errors that you can use?

dbwiddis commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+                        } else {
+                            DetectorResult result = DetectorResult
+                                .failedValidation(detector.getIndices().get(0), "Non-blocking warning: " + errorMessage);
+                            listener.onResponse((T) result.toJson());
```
If `detector.getIndices().get(0)` throws an exception the listener will never be completed. Probably need some exception handling here.

kaituo commented on src/main/resources/org/opensearch/agent/tools/CreateAnomalyDetectorEnhancedPrompt.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "CLAUDE": "\n\nHuman: Analyze this index and suggest ONE anomaly detector for operational monitoring that generates clear, actionable alerts.\n\nIndex: ${indexInfo.indexName}\nMapping: ${indexInfo.indexMapping}\nAvailable date fields: ${dateFields}\n\nCORE PRINCIPLE:\nCreate detectors that will find anomalies on operational issues teams can immediately investigate and resolve.\nFocus on 1-2 REL ...
```
>e.g., user_agent.keyword, message.keyword

What if I want to monitor the message volume and the number of agents running?

kaituo replied:
>category_field=field_name_or_empty

Should we tell LLM what happens when there is no categorical field?

kaituo replied:
>interval=minutes

Should we tell LLM what happens when there are no suitable interval found?

kaituo replied:
>can immediately investigate and resolve

Can you define  what is the criteria by which the team can investigate and resolve?

kaituo commented on src/main/resources/org/opensearch/agent/tools/CreateAnomalyDetectorEnhancedPrompt.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "CLAUDE": "\n\nHuman: Analyze this index and suggest ONE anomaly detector for operational monitoring that generates clear, actionable alerts.\n\nIndex: ${indexInfo.indexName}\nMapping: ${indexInfo.indexMapping}\nAvailable date fields: ${dateFields}\n\nCORE PRINCIPLE:\nCreate detectors that will find anomalies on operational issues teams can immediately investigate and resolve.\nFocus on 1-2 REL ...
+  "OPENAI": "Analyze this index and suggest ONE anomaly detector for operational monitoring that generates clear, actionable alerts.\n\nIndex: ${indexInfo.indexName}\nMapping: ${indexInfo.indexMapping}\nAvailable date fields: ${dateFields}\n\nCORE PRINCIPLE:\nCreate detectors that will find anomalies on operational issues teams can immediately investigate and resolve.\nFocus on 1-2 RELATED fields ...
```
It is hard to maintain two prompts with almost identical content. Can you try to generate the claude one programmatically by adding extra information?

kaituo commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorTool.java:
```diff
@@ -335,21 +336,8 @@ private Map<String, String> enrichParameters(Map<String, String> parameters) {
-    }
-
     @SuppressWarnings("unchecked")
+    **/
```
@SuppressWarnings should be outside of comment block. This may caused CI compile failure.

kaituo commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+        TimeConfiguration newWindowDelay = originalDetector.getWindowDelay();
+        Integer newHistoryIntervals = originalDetector.getHistoryIntervals();
+        if (response.getInterval() != null) {
+            newInterval = response.getInterval();
```
Would this override LLM's interval? If yes, is this on purpose? why do we need LLM's interval suggestion?

kaituo reviewed (commented):
partial review

kaituo commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+
+        // Check if this is a sparse data issue with unreasonably high suggested interval
+        boolean isUnreasonableInterval = validationError.contains("interval")
+            && (validationError.contains("240")
```
Can we just say the interval is larger than 240, isUnreasonableInterval is true? Suggest API can give a larger number than 1440.

kaituo commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+                listener.onResponse(bestDateField);
+            })
+            .exceptionally(e -> {
+                futures.forEach(FutureUtils::cancel);
```
The futures may have already completed.

kaituo commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+                result = DetectorResult.failedStart(indexName, null, errorMessage);
+                break;
+            default:
+                throw new IllegalArgumentException("Unknown status: " + status);
```
Throw listener failure? Otherwise, this leaves the listener hanging.

kaituo commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+        if (parameters.containsKey("input")) {
+            String inputStr = parameters.get("input");
+            if (inputStr != null && inputStr.trim().startsWith("[")) {
+                parameters.put("input", "{\"indices\": " + inputStr + "}");
```
Should we sanitize user input? Is it possible we may have prompt injection from malicious users?

kaituo commented on src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java:
```diff
@@ -0,0 +1,1402 @@
+            null,
+            null,
+            null,
+            new IntervalTimeConfiguration(intervalMinutes, ChronoUnit.MINUTES),
```
Frequency should be a random number between [2*interval, 24h] and a multiple of interval. Ideally, we can just put 24h as we don't require real time monitoring. But to have some jitter, we apply some random number. Otherwise, we risk breaking down the domain by running too many concurrent queries at the same time.

amitgalitz committed:
add index insight integration with fallback

Signed-off-by: Amit Galitzky <amgalitz@amazon.com>

amitgalitz committed:
adding otel built in detectors

Signed-off-by: Amit Galitzky <amgalitz@amazon.com>

amitgalitz committed:
code cleanup addressing several comments

Signed-off-by: Amit Galitzky <amgalitz@amazon.com>

amitgalitz committed:
adding sequential detector creation

Signed-off-by: Amit Galitzky <amgalitz@amazon.com>

amitgalitz committed:
Add prompt filter support, custom result index, and sequential multi-detector tests

Signed-off-by: Amit Galitzky <amgalitz@amazon.com>

amitgalitz committed:
Production hardening: frequency jitter, null field filter, retry memory

- Jittered detector frequency (random multiple of interval, 2x-24h) to
  prevent concurrent query storms across detectors
- Filter null fields via sampler aggregation (100K docs, 0.1% threshold)
  before sending mapping to LLM, same approach as Index Insight
- Thread retry memory through detector and model validation so LLM does
  not repeat failed configurations
- Append aggregation method to feature names to prevent collisions
- Extract shared buildRetryContext helper, remove dead generateAndParseConfig

Signed-off-by: Amit Galitzky <amgalitz@amazon.com>

amitgalitz committed:
Flexible LLM parser, OTel validation, input guards, prompt consolidation, code cleanup

Signed-off-by: Amit Galitzky <amgalitz@amazon.com>
